File: cowemb/cowemb.hs
module Main 
  (
    main
  ) where

import Data.Array (listArray, (!))
import Control.Monad (liftM)

data Line = Line Int Int Int

-- Helper functions
count :: (a -> Bool) -> [a] -> Int
count p xs = length $ filter p xs

-- Solve
solve :: Int -> [Line] -> Int
solve r lines = 
    let n = length lines
        a = listArray (0, n-1) lines in
    count (uncurry $ meet r) [(a!i, a!j) | i <- [0..n-1], j <- [0..i-1]]

-- Check if two lines meet within raduis r of the origin.
meet :: Int -> Line -> Line -> Bool
meet r line1 line2 = 
    let Line a b e = line1
        Line c d f = line1 in  
    (e*d - b*f)^2 + (a*f - e*c)^2 <= r^2 * (d*b - b*c)^2

-- I/O procedure
main :: IO ()
main = do  
    (n, r) <- readLine parInts
    lines <- sequence . replicate n $ readLine parThread
    print $ solve r lines 
      where 
        readLine par = return . par . words =<< getLine
        parInts (n:r:_) = (read n, read r)
        parThread (a:b:c:_) = Line (read a) (read b) (-(read c))

File: cowemb/cowemb.cc
#include <iostream>
#include <vector>

using namespace std;

struct Line {
    Line(int a, int b, int c)
      : a(a),
        b(b),
        c(c)
    {}
    int a, b, c;
};

bool meet(int r, Line line1, Line line2)
{
    int a = line1.a;
    int b = line1.b;
    int e = line1.c;

    int c = line2.a;
    int d = line2.b;
    int f = line2.c;

    long int foo = e*d - b*f;
    long int bar = a*f - c*e;
    long int baz = a*d - b*c;

    return foo*foo + bar*bar <= r*r * baz*baz;
}

int main(int, const char**)
{
    int n, r;
    cin >> n >> r;

    vector<Line> lines;
    for (int i = 0; i < n; ++i) {
        int a, b, c;
        cin >> a >> b >> c;
        lines.push_back(Line(a, b, -c));
    }

    long int count = 0;
    for (int i = 0; i < n; ++i)
        for (int j = i + 1; j < n; ++j) 
            if (meet(r, lines[i], lines[j]))
                ++count;
    
    cout << count << endl;

    return 0;
}

